lib/Command/RecreateMasterKey.php

```diff
@@ -0,0 +1,338 @@
+								ISecureRandom $secureRandom, IL10N $l, ILogger $logger) {
+
+		parent::__construct();
+		$this->userManager = $userManager;
+		$this->rootView = $rootView;
+		$this->keyManager = $keyManager;
+		$this->util = $util;
+		$this->IAppManager = $IAppManager;
+		$this->appConfig = $appConfig;
+		$this->config = $config;
+		$this->session = $session;
+		$this->questionHelper = $questionHelper;
+		$this->userSetup = $userSetup;
+		$this->mailer = $mailer;
+		$this->secureRandom = $secureRandom;
+		$this->l = $l;
+		$this->logger = $logger;
+	}
+
+	protected function configure() {
+		parent::configure();
+
+		$this
+			->setName('encryption:recreate-master-key')
+			->setDescription('Replace existing master key with new one. Encrypt the file system with newly created master key')
+		;
+
+		$this->addOption(
+			'yes',
+			'y',
+			InputOption::VALUE_NONE,
+			'Answer yes to all questions'
+		);
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output) {
+		$yes = $input->getOption('yes');
+		if ($this->util->isMasterKeyEnabled()) {
+			$question = new ConfirmationQuestion(
+				'Warning: Inorder to re-create master key, the entire ownCloud filesystem will be decrypted and then encrypted using new master key.'
+				. ' Do you want to continue? (y/n)', false);
+			if ($yes || $this->questionHelper->ask($input, $output, $question)) {
+
+				$output->writeln("Decryption started\n");
+				$progress = new ProgressBar($output);
+				$progress->start();
+				$progress->setMessage("Decryption progress...");
+				$progress->advance();
+
+				$this->decryptAllUsersFiles($progress, $output);
+				$progress->finish();
+
+				if (empty($this->failed)) {
+
+					$this->IAppManager->disableApp('encryption');
+
+					//Delete the files_encryption dir
+					$this->rootView->deleteAll('files_encryption');
+
+					$this->appConfig->setValue('core', 'encryption_enabled', 'no');
+					$this->appConfig->deleteKey('encryption', 'useMasterKey');
+					$this->appConfig->deleteKey('encryption', 'masterKeyId');
+					$this->appConfig->deleteKey('encryption', 'recoveryKeyId');
+					$this->appConfig->deleteKey('encryption', 'publicShareKeyId');
+					$this->appConfig->deleteKey('files_encryption', 'installed_version');
+
+				}
+				$output->writeln("\nDecryption completed\n");
+
+				//Reencrypt again
+				$this->IAppManager->enableApp('encryption');
+				$this->appConfig->setValue('core', 'encryption_enabled', 'yes');
+				$this->appConfig->setValue('encryption', 'enabled', 'yes');
+				$output->writeln("Encryption started\n");
+
+				$output->writeln("Waiting for creating new masterkey\n");
+
+				$this->keyManager->setPublicShareKeyIDAndMasterKeyId();
+
+				$output->writeln("New masterkey created successfully\n");
+
+				$this->appConfig->setValue('encryption', 'enabled', 'yes');
+				$this->appConfig->setValue('encryption', 'useMasterKey', '1');
+
+				$this->keyManager->validateShareKey();
```

Comment:
what happens if this fails ? will it throw an exception, should we catch it ?

Reply:
It would try to create keys if it doesn't exist. If key exist nothing will be done. Same is for `validateMasterKey`